### ISSUE TITLE
PHP Fatal error:  Call to a member function handshake() on a non-object

### ISSUE THREAD

**Issue by abienvenu** [ratchetphp/Ratchet#238]
My 0.3.2 Ratchet server has been running smoothly for months on PHP 5.5.10, with hundreds of people connected to it. And last night, it crashed :

    [02-Oct-2014 03:32:55 Europe/Paris] PHP Notice:  Undefined property: stdClass::$version in /symfony/vendor/cboden/ratchet/src/Ratchet/WebSocket/WsServer.php on line 114
    [02-Oct-2014 03:32:55 Europe/Paris] PHP Fatal error:  Call to a member function handshake() on a non-object in /symfony/vendor/cboden/ratchet/src/Ratchet/WebSocket/WsServer.php on line 114

It happened in the middle of the night, so the client was probably a broken bot or some hacker.

Just in case, here is my main loop :
```php
		$server = new Server();
		$wsServer = new WsServer($server);
		$httpServer = new HttpServer($wsServer);
		$loop = IoServer::factory($httpServer, $port);
		$loop->run();
```

And here is my Server class :
```php
use Ratchet\MessageComponentInterface;
use Ratchet\ConnectionInterface;

class Server implements MessageComponentInterface
{
	protected $clients;

	public function __construct()
	{
		$this->clients   = new \SplObjectStorage;
	}

	public function onOpen(ConnectionInterface $connection)
	{
		$this->clients->attach($connection);
	}

	public function onMessage(ConnectionInterface $connection, $message)
	{
		// Do stuff

		if ($broadcastOk) {
			foreach ($this->clients as $client) {
				$client->send(json_encode($data));
			}
		}
		else {
			$connection->close();
		}
	}

	public function onClose(ConnectionInterface $connection)
	{
		$this->clients->detach($connection);
	}

	public function onError(ConnectionInterface $connection, \Exception $e)
	{
		$this->clients->detach($connection);
		$connection->close();
	}
}
```

Any ideas about the cause of this crash ?

**Comment by mart1** [comment]
I have your heisenbug with my code: 

PHP Notice:  Undefined property: stdClass::$version in /root/MyApp/vendor/cboden/ratchet/src/Ratchet/WebSocket/WsServer.php on line 109
PHP Fatal error:  Call to a member function handshake() on a non-object in /root/MyApp/vendor/cboden/ratchet/src/Ratchet/WebSocket/WsServer.php on line 109

```php

namespace Crm;
use Ratchet\MessageComponentInterface;
use Ratchet\ConnectionInterface;

class Dispatch implements MessageComponentInterface {
	protected $clients;
	protected $pamiConn;
	public function __construct() {
		$this->clients = array();
		$this->pamiConn = null;
	}
	
	public function onOpen(ConnectionInterface $conn) {
		// Store the new connection to send messages to later
		//$this->clients->attach ( $conn );
		echo "New connection! ({$conn->resourceId})\n";
	}
	
	public function onMessage(ConnectionInterface $from, $msg) {
		echo "onMessage\n";
		print_r($msg);
		echo "\n";
		$json = json_decode ( $msg );
		if (isset ( $json->sessid )) {
			$sessid = $json->sessid;
			echo "$sessid\n";
			// code special pour le PAMI
			if ($sessid == "P@mI") {
				echo "PAMI connected\n";
				$this->pamiConn = $from;
				$this->clients["PAMI"]=array("from"=>$from);
			} else {
				// on ne recoit que la session
				// on recupere le device pour ce sessid
				$opts = array ('http' => array ('method' => 'POST', 'header' => 'Content-type: application/x-www-form-urlencoded', 'content' =>  http_build_query(array ('session' => $sessid ) ) ) );
				$context = stream_context_create ( $opts );
				///$r = new HttpRequest ( 'https://produit.epixelic.net/api/device', HttpRequest::METH_POST );
				
				//$r->addPostFields ( );
			
				
				try {
					$res = file_get_contents('https://produit.epixelic.net/api/device.json', false, $context);
					echo "$res\n";
					$json = json_decode ( $res );
					if (isset ( $json->result->message ) && $json->result->message == 'OK') {
						$mac = json_decode ( $res )->result->device;

						if(array_key_exists($mac,$this->clients)){
							echo $mac . " already connected (écrasement)\n";
							$this->clients[$mac] = array ("sessid" => $sessid, "from" => $from ) ;
						}else{
							$this->clients[$mac] = array ("sessid" => $sessid, "from" => $from ) ;
							echo $mac . " connected\n";
						}

						
					}
					if(!is_null($this->pamiConn)){
						// TODO:déclaration du téléphone auprès de PAMI
						$this->pamiConn->send (json_encode(array("action"=>"add","mac"=>$mac)));
					}else
					{
						echo "PAMI n'est pas connecte\n";
					}
				} catch ( HttpException $ex ) {
					echo $ex;
				}
			}
		
		} else if (! is_null ( $this->pamiConn ) && $this->pamiConn == $from) { // pas de sessid alors c'est PAMI
			$json = json_decode ( $msg );
			//$json = json_decode(json_encode ( array ("number" => "4545", "mac" => "5454" ) ));
			// TODO: un message de PAMI a relayer a un telephone
			if (isset ( $json->mac ) && isset ( $json->number )) {
				
				if(isset($this->clients [$json->mac])){

					$this->clients [$json->mac] ["from"]->send ('{"result":{"message":"OK","callernumber":"'.$json->number.'"}}');


				}else{
					echo "Telephone non enregistre";
				}
			} else {
				echo "Données manquantes";
			}
		} else {
			echo "Demande inconnue";
		}
	
		// on récupere le device par rapport au nic
	

	}
	public function close(){
		foreach($this->clients as $data){
			$data["from"]->close();
		}
        }
	
	public function onClose(ConnectionInterface $conn) {
		// The connection is closed, remove it, as we can no longer send it messages
		
		foreach($this->clients as $mac=>$data){
			if($data["from"]==$conn){
				echo  $mac ." disconnected\n";
				unset($this->clients[ $mac ]);
				
				if(!is_null($this->pamiConn))$this->pamiConn->send (json_encode(array("action"=>"remove","mac"=>$mac)));
				break;
			}
		}
		 
		echo "Connection {$conn->resourceId} has disconnected\n";
	}
	
	public function onError(ConnectionInterface $conn,\Exception $e) {
		echo "An error has occurred: {$e->getMessage()}\n";
		
		$conn->close ();
	}
}
```

**Comment by cboden** [comment]
@mart1 Are you able to provide steps to reproduce the problem? What browser were you using when the error happened?

@abienvenu @mart1 @RynnHeldeD I made PR #337. Can you guys test that out and see if the error continues? I think I see *how* the error has occurred but I don't know *what* is causing it to happen. My best guess is a browser is doing something it shouldn't be doing.

**Comment by RynnHeldeD** [comment]
@cboden Thanks for the fix, I'll try it as soon as I'll understand how to implement it.
As I installed Ratchet in my project using Composer, I suppose I just have to replicate the changes you made in your PR #337, then test my project again until the problem happens again (or never if it really fixes it) ?

**Comment by cboden** [comment]
@RynnHeldeD Update your composer file from Ratchet's version to look like this (and run composer update):

> "cboden/ratchet": "dev-handshake-fix"

**Comment by RynnHeldeD** [comment]
Ok, done !
I'll tell you if there is some news about this strange bug.

2015-07-15 22:38 GMT+02:00 Chris Boden <notifications@github.com>:

**Comment by cboden** [comment]
Hopefully fixed in #337. Please report if this issue still persists.
